crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,335 @@
+        Self {
+            storage_arc: Arc::new(0.into(), 0),
+            uptime_0_to_1: 0.0,
+        }
+    }
+}
+
+impl RValuePeerRecord {
+    /// what arc is this agent claiming to store
+    pub fn storage_arc(mut self, storage_arc: Arc) -> Self {
+        self.storage_arc = storage_arc;
+        self
+    }
+
+    /// what is this agent's uptime
+    /// - 0 = never reachable
+    /// - 1 = always reachable
+    pub fn uptime_0_to_1(mut self, uptime_0_to_1: f64) -> Self {
+        self.uptime_0_to_1 = uptime_0_to_1;
+        self
+    }
+}
+
+/// builder for a set of peers we know about
+/// this should generally either be our entire storage arc of peer info
+/// or, if it is huge, small sample arc
+pub struct RValuePeerRecordSet {
+    arc_of_included_peer_records: Arc,
+    peer_records: Vec<RValuePeerRecord>,
+}
+
+impl Default for RValuePeerRecordSet {
+    fn default() -> Self {
+        Self {
+            arc_of_included_peer_records: Arc::new(0.into(), 0),
+            peer_records: Vec::new(),
+        }
+    }
+}
+
+impl RValuePeerRecordSet {
+    /// the arc we are representing with this PeerRecordSet
+    pub fn arc_of_included_peer_records(mut self, arc_of_included_peer_records: Arc) -> Self {
+        self.arc_of_included_peer_records = arc_of_included_peer_records;
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record does not fit within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_avg_coverage = 0.0_f64;
+
+    for (i, record) in peer_record_set.peer_records.iter().enumerate() {
+        // what is the total coverage factor for this record?
+        let coverage =
+            (record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64) * record.uptime_0_to_1;
+
+        // mix it into our running average
+        running_avg_coverage = (running_avg_coverage * i as f64 + coverage) / (i as f64 + 1.0);
+    }
+
+    // given our known arc, estimate the total node count
+    let est_node_count = peer_record_set.peer_records.len() as f64
+        * (1.0
+            / (peer_record_set.arc_of_included_peer_records.length() as f64
+                / ARC_LENGTH_MAX as f64));
```

Comment:
Couldn't this just be
`* (ARC_LENGTH_MAX as f64 / peer_record_set.arc_of_included_peer_records.length() as f64`
?

To me, it's just clearer to use the inverse ratio total DHT arc vs. our storage arc to scale up. I found myself doing some mental gymnastics to realize what was being achieved here with double division.

Reply:
lol : )